Stack IR

kripken opened the pull request:
This adds a new IR, "Stack IR". This represents wasm at a very low level, as a simple stream of instructions, basically the same as wasm's binary format. This is unlike Binaryen IR which is structured and in a tree format.

This gives some small wins on binary sizes, less than 1% in most cases, usually 0.25-0.50% or so. That's not much by itself, but looking forward this prepares us for multi-value, which we really need an IR like this to be able to optimize well. Also, it's possible there is more we can do already - currently there are just a few stack IR optimizations implemented,

 * DCE
 * local2stack - check if a `set_local/get_local` pair can be removed, which keeps the set's value on the stack, which if the stars align it can be popped instead of the get.
 * Block removal - remove any blocks with no branches, as they are valid in wasm binary format.

Implementation-wise, the IR is defined in `wasm-stack.h`. A new `StackInst` is defined, representing a single instruction. Most are simple reflections of Binaryen IR (an add, a load, etc.), and just pointers to them. Control flow constructs are expanded into multiple instructions, like a block turns into a block begin and end, and we may also emit extra unreachables to handle the fact Binaryen IR has unreachable blocks/ifs/loops but wasm does not. Overall, all the Binaryen IR differences with wasm vanish on the way to stack IR.

Where this IR lives: Each Function now has a `unique_ptr` to stack IR, that is, a function may have stack IR alongside the main IR. If the stack IR is present, we write it out during binary writing; if not, we do the same binaryen IR => wasm binary process as before (this PR should not affect speed there). This design lets us use normal Passes on stack IR, in particular this PR defines 3 passes:

 * Generate stack IR
 * Optimize stack IR (might be worth splitting out into separate passes eventually)
 * Print stack IR for debugging purposes

Having these as normal passes is convenient as then they can run in parallel across functions and all the other conveniences of our current Pass system. However, a downside of keeping the second IR as an option on Functions, and using normal Passes to operate on it, means that we may get out of sync: if you generate stack IR, then modify binaryen IR, then the stack IR may no longer be valid (for example, maybe you removed locals or modified instructions in place etc.). To avoid that, Passes now define if they modify Binaryen IR or not; if they do, we throw away the stack IR.

Miscellaneous notes:
 * Just writing Stack IR, then writing to binary - no optimizations - is 20% slower than going directly to binary, which is one reason why we still support direct writing. This does lead to some "fun" C++ template code to make that convenient: there is a single StackWriter class, templated over the "mode", which is either Binaryen2Binary (direct writing), Binaryen2Stack, or Stack2Binary. This avoids a lot of boilerplate as the 3 modes share a lot of code in overlapping ways.
 * Stack IR does not support source maps / debug info. We just don't use that IR if debug info is present.
 * A tiny text format comment (if emitting non-minified text) indicates stack IR is present, if it is (`(; has Stack IR ;)`). This may help with debugging, just in case people forget. There is also a pass to print out the stack IR for debug purposes, as mentioned above.
 * The sieve binaryen.js test was actually not validating all along - these new opts broke it in a more noticeable manner. Fixed.
 * I ran the emscripten test suite + fuzzing. Will do some more testing too.

dcodeIO commented:
Not sure if this is heading into this direction, but it would be a nice bonus if tooling based upon Binaryen could provide output of an entire module in stacky text format as an option some day as well :)

kripken commented:
Yeah, actually the `--print-stack-ir` pass almost does what you want there. The differences are

 * It doesn't print the immediate values of an instruction, so it prints `load` instead of `load offset=4` etc. We could refactor the `--print` pass to move the immediate value printing into helper functions (so `visitLoad` would call `printLoadContents` or such) etc., then call that from the stack IR printing code too.
 * It doesn't add indentation yet. Trivial to add.
 * It prints a numeric index for each instruction. I found that nice for debugging, but we can remove it.

So if proper stack IR text output would be useful, that should be easy to do.

kripken commented:
* Added extra checks in pass-debug mode, to verify that if stack IR should have been thrown out, it was. This should help avoid any confusion with the IR being invalid.
 * Added a comment about the possible future of stack IR as the main IR, depending on optimization results, following some discussion earlier today.

jgravelle-google commented on src/passes/StackIR.cpp:
```diff
@@ -0,0 +1,391 @@
+  bool modifiesBinaryenIR() override { return false; }
+
+  void doWalkFunction(Function* func) {
+    if (!func->stackIR) {
```
Should this check live in StackIROptimizer itself? Or should `StackIROptimizer(...).run()` never be a no-op, at which point we should maybe add an assert to `StackIROptimizer` itself.

kripken replied:
It could be in both places I guess. I like the idea that `StackIROptimizer` operates on valid input, containing stack IR, and the pass is valid to run if there is or isn't. Yeah, makes sense to add an assert inside StackIROptimizer, I'll add that.

kripken commented:
(Added a few commits for fuzzer issues.)

jgravelle-google commented on src/passes/StackIR.cpp:
```diff
@@ -0,0 +1,393 @@
+    // We also maintain a stack of values vectors for control flow,
+    // saving the stack as we enter and restoring it when we exit.
+    std::vector<std::vector<Index>> savedValues;
+#ifdef STACK_OPT_DEBUG
```
Should this be a `#define`? Probably for now I guess, because we won't want to use this unless developing the stack optimizer. Can change it later if we need to, for now probably best to keep it simple.

jgravelle-google commented on src/passes/StackIR.cpp:
```diff
@@ -0,0 +1,393 @@
+          // can reach the set.
+          if (values.size() > 0) {
+            Index j = values.size() - 1;
+            while (1) {
```
Is this any less weird?
```
for (Index j1 = values.size(); j1 != 0; --j1) {
  Index j = j1 - 1;
```
The `while (1)` is kinda weird... but so is the off-by-one for loop :/

kripken replied:
Yeah, it's weird either way I think... it's just awkward to iterate down to 0 on an unsigned number.

kripken commented:
In the `stack4` branch I did a quick experiment with refactoring the printing code to emit valid `wat` text when we print stack IR. This turned out to be surprisingly easy, and I did some fuzzing using wabt to read those `wat` files, and can't find any bugs there. I think that is further support for the stack representation being correct, and also also that this approach can get us quickly to being able to emit valid wat (which has been requested, see e.g. https://github.com/AssemblyScript/assemblyscript/issues/179#issuecomment-408302345 ).
